### PR TITLE
I am going to build my own function with move semantics and polymorphism

### PR DESCRIPTION
**Pull request by eggrobin**

### CODE REVIEW

File: base/function.hpp [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+class Functor {
+ public:
+  virtual ~Functor() = default;
+  virtual Result call(Args&&...) = 0;
+};
+
+template<typename F, typename Result, typename... Args>
+class ConcreteFunctor final : public Functor<Result, Args...> {
+ public:
+  explicit ConcreteFunctor(F Functor);
```
Comment (pleroy) [inline base/function.hpp]: `F functor`

File: base/function.hpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  Result call(Args&&... args) final;
+ private:
+  F functor_;
+};
+
+template<typename Signature>
+struct function;
```
Comment (pleroy) [inline base/function.hpp]: `class`

File: base/function_body.hpp [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+#include <utility>
+
+#include "base/function.hpp"
+
+namespace principia {
+namespace base {
+namespace internal_function {
+
+template<typename F, typename Result, typename... Args>
+ConcreteFunctor<F, Result, Args...>::ConcreteFunctor(F Functor)
```
Comment (pleroy) [inline base/function_body.hpp]: `F functor`

File: base/function.hpp [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+namespace principia {
+namespace base {
+namespace internal_function {
+
+template<typename Result, typename... Args>
+class Functor {
+ public:
+  virtual ~Functor() = default;
+  virtual Result call(Args&&...) = 0;
```
Comment (pleroy) [inline base/function.hpp]: `Call`, maybe?

File: base/function_test.cpp [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+TEST_F(FunctionTest, MovableFunction) {
+  int observe_i = 0;
+  function<void()> λ = [i = std::make_unique<int>(), &observe_i](){
+    ++*i;
+    observe_i = *i;
+  };
+  function<void()> f = std::move(λ);
+  f();
+  EXPECT_THAT(observe_i, Eq(1));
```
Comment (pleroy) [inline base/function_test.cpp]: `EXPECT_EQ` (several places)

File: base/function_body.hpp (hunk 3 above)
Comment (eggrobin) [inline base/function_body.hpp]: Done

File: base/function.hpp (hunk 2 above)
Comment (eggrobin) [inline base/function.hpp]: Done

File: base/function.hpp (hunk 1 above)
Comment (eggrobin) [inline base/function.hpp]: Done

File: base/function.hpp (hunk 4 above)
Comment (eggrobin) [inline base/function.hpp]: Done

File: base/function_test.cpp (hunk 5 above)
Comment (eggrobin) [inline base/function_test.cpp]: no

**Pull request merged**